[LLVM][Clang][AArch64] Implement AArch64 build attributes

sivan-shani opened the pull request:
- Added support for AArch64-specific build attributes.
- Print AArch64 build attributes to assembly.
- Parse AArch64 build attributes from assembly.
- Emit AArch64 build attributes to ELF.

Specification: https://github.com/ARM-software/abi-aa/blob/ada00cc8e04f421cce657e8d9f3a439c69437cf4/buildattr64/buildattr64.rst

sivan-shani committed:
[LLVM][Clang][AArch64] Implement AArch64 build attributes

- Added support for AArch64-specific build attributes.
- Print AArch64 build attributes to assembly.
- Emit AArch64 build attributes to ELF.

Specification: https://github.com/ARM-software/abi-aa/blob/654a64cbac041fc3bff617800998d40b5068f592/buildattr64/buildattr64.rst#aeabi-feature-and-bits-subsection

sivan-shani committed:
Formatting

ostannard commented on llvm/include/llvm/MC/MCELFStreamer.h:
```diff
@@ -107,25 +108,51 @@ class MCELFStreamer : public MCObjectStreamer {
+    // [<uint32: subsection-length> NTBS: vendor-name <bytes: vendor-data>]*
+    StringRef Vendor;
+    // <uint8: optional> <uint8: parameter type> <attribute>*
+    unsigned IsMandatory; // SubsectionMandatory::REQUIRED (0),
```
You have added enums for these in ARMBuildAttributes.h, should they be used here?

sivan-shani replied:
Yes, that is a good idea.
Changing the relevant locations.

ostannard commented on llvm/lib/MC/MCELFStreamer.cpp:
```diff
@@ -783,6 +786,65 @@ void MCELFStreamer::createAttributesSection(
   AttrsVec.clear();
 }
 
+void MCELFStreamer::createAttributesSection(
```
If I've understood correctly, AArch64 uses a slightly different format to AArch32, RISC-V and Hexagon, so the function names should probably make this clear, instead of using overloading to select between the two formats.

sivan-shani replied:
Does it has to be the functions names, or could it be explained in comments?

ostannard commented on llvm/include/llvm/Support/ARMBuildAttributes.h:
```diff
@@ -21,10 +21,49 @@
 const TagNameMap &getARMAttributeTags();
 
+/// AArch64 build attributes vendors (=subsection name)
+enum Vendor : unsigned { AEBI_FEATURE_AND_BITS = 0, AEBI_PAUTHABI = 1 };
```
s/AEBI/AEABI/

ostannard commented on llvm/lib/Target/ARM/MCTargetDesc/ARMELFStreamer.cpp:
```diff
@@ -793,20 +793,23 @@ void ARMTargetELFStreamer::switchVendor(StringRef Vendor) {
   getStreamer().setAttributeItem(Attribute, Value,
-                                 /* OverwriteExisting= */ true);
+                                 /* OverwriteExisting= */ true,
+                                 getStreamer().Contents);
```
I don't like this API change, requiring non-AArch64 targets to pass a member variable of the stream back into the streamer every time. What about adding a new function to emit an attribute in a specific sub-section?

sivan-shani replied:
Emitting into a specific subsection is not complicated.
All that is required is 1. first, create the subsection and add it to the subsections vector (AArch64TargetELFStreamer::emitSubsection) 2. then, use AArch64TargetELFStreamer::emitAttribute which will emit it into the correct subsection. 

Providing the data structure as an argument to the function rather then using the 'this' pointer (= the data structure that is in the class) data structure is, arguably, the more 'correct' way to construct a function in principal. Therefore it did not seems as a concern to change it. It might seems as a preferable solution to adding more functions to the base class.

Let me know what you think, no objection to changing it if you still think it is the better approach.

sivan-shani replied:
Clarification: 
The part about the mechanism m for emitting attributes refer to ELF.
Asm attributes need to be emitted in order since they are being printed directly.

ostannard replied:
Now that you've added the concept of the "active subsection", is this still needed, or could this be changed back to always emit into the active subsection?

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp:
```diff
@@ -294,6 +379,58 @@ AArch64ELFStreamer &AArch64TargetELFStreamer::getStreamer() {
+  // If exists, return.
+  for (MCELFStreamer::AttributeSubSection &SubSection : AttributeSubSections) {
+    if (SubSection.Vendor == VendorAsStr) {
+      llvm_unreachable("AArch64 build attributes subsection already exists");
```
Is it not valid to switch back to a previous subsection, for example from user-written assembly directives?

sivan-shani replied:
It is valid and possible to emit attribute to any subsection at any time, this achieved by the usage of a vector to hold the subsections.
There is no need to mechanically 'switch' between subsections. After a subsection has been created, it is always available and it is always possible to emit an attribute into it.

For example, user can first creates all the subsections that he wants, them add attribute to any of them at any time.

This function creates a new subsection and add it to the vector. 
The check here is for the case of creating the same subsection twice.

sivan-shani replied:
Clarification: 
> It is valid and possible to emit attribute to any subsection at any time
Referring to the ELF mechanism. Asm attributes need to be emitted in order since they are being printed directly.

sivan-shani replied:
This design does not require switching between sections in order to emit an ELF attribute to a specific section. As long as the required subsection has been created, an attribute can be emitted into it.

sivan-shani replied:
It is valid. Programmatically, it is not required, attribute will be added to the correct subsection. When writing assembly it is done via adding the correct subsection header.

ostannard commented on llvm/lib/Support/ARMBuildAttrs.cpp:
```diff
@@ -67,7 +69,56 @@ static const TagNameItem tagData[] = {
-constexpr TagNameMap ARMAttributeTags{tagData};
-const TagNameMap &llvm::ARMBuildAttrs::getARMAttributeTags() {
-  return ARMAttributeTags;
+const TagNameItem AVAttributeTags[] = {
```
These are downstream changes we have to work with armlink, and shouldn't be included here.

ostannard reviewed (commented):
This needs tests.

What about the assembly parser for the new directives?

asl commented:
Tagging @kovdan01

sivan-shani commented:
> This needs tests.
Will fix the comments and commit, then will commit tests.
> What about the assembly parser for the new directives?
Work divided to emitting/parsing. This change cover emitting Asm and ELF, will add parsers later.

sivan-shani committed:
Fixing according to some comments:
- make use of the enums
- s/AEBI/AEABI/
- change function name (createAttributesSection)
- remove AVAttribute code

sivan-shani committed:
Add test files

ostannard commented:
The code changes are all in llvm (not clang), so the tests need to be there too, using `llvm-mc` and `llc` instead of `clang`.

I'd prefer to see the code for the assembly parser before reviewing this, because that will probably affect the API design of the streamer, since it will need to handle switching back and forth between subsections in ways llc doesn't need.

sivan-shani committed:
Add Asm parsing
Add some fixes

sivan-shani committed:
Formatting

sivan-shani committed:
add llvm test files

sivan-shani commented:
- added llvm tests
- added assembly parser code

sivan-shani committed:
Merge branch 'main' into AArch64Buildattributes

sivan-shani committed:
format

sivan-shani committed:
fix small bug

sivan-shani committed:
Fix a bug

sivan-shani committed:
format

sivan-shani committed:
- add llvm-mc test files
- fix bug

ostannard commented on llvm/include/llvm/MC/MCELFStreamer.h:
```diff
@@ -105,27 +107,55 @@ class MCELFStreamer : public MCObjectStreamer {
+    // [<uint32: subsection-length> NTBS: vendor-name <bytes: vendor-data>]*
+    StringRef VendorName;
+    // <uint8: optional> <uint8: parameter type> <attribute>*
+    ARMBuildAttrs::SubsectionOptional IsOptional;
```
AArch64-specific data types shouldn't be used in target-independent headers. Either `SubsectionOptional` and `SubsectionType` should be moved to a target-independent header, or the whole `AttributeSubSection` type should be moved down to an AArch64 file.

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp:
```diff
@@ -148,13 +151,133 @@ class AArch64TargetAsmStreamer : public AArch64TargetStreamer {
+    }
+    case ARMBuildAttrs::AEABI_FEATURE_AND_BITS:
+      switch (Tag) {
+      default:
```
We need to allow emitting any attribute by number, so that it is possible to use future attributes without upgrading the assembler.

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64TargetStreamer.cpp:
```diff
@@ -151,3 +151,81 @@ llvm::createAArch64ObjectTargetStreamer(MCStreamer &S,
   return new AArch64TargetStreamer(S);
 }
+
+void AArch64TargetStreamer::emitSubsection(
```
The term "subsection" already has a meaning in MC, so this should be `emitAtributesSubsection`. Same comment applies to other names.

ostannard reviewed (commented):
It looks like this is still missing the ability to emit build attributes by number, for future attributes the assembler does not know about.

There is also still a lot of new code here not covered by tests:
* Tests for pauthabi platform/version in codegen
* Tests for assembly parser errors
* Tests for attributes with NTBS values

sivan-shani commented on llvm/lib/MC/MCELFStreamer.cpp:
```diff
@@ -783,6 +786,67 @@ void MCELFStreamer::createAttributesSection(
   AttrsVec.clear();
 }
 
+void MCELFStreamer::createAArch64AttributesSection(
```
Made available for generic usage.

sivan-shani committed:
- Fixed according to comments above
- Fixed some errors
- Ability to add attributes numerically added
- Tests for pauthabi platform/version in codegen
- Tests for assembly parser errors
- Not added: Tests for attributes with NTBS values - currently all attributes takes only ULEB128 as values.

sivan-shani commented:
- Fixed according to comments above
- Fixed some errors
- Ability to add attributes numerically added
- Tests for pauthabi platform/version in codegen
- Tests for assembly parser errors
- Not added: Tests for attributes with NTBS values - currently all attributes takes only ULEB128 as values.

sivan-shani committed:
- remove unused variables and includes

ostannard commented:
> Not added: Tests for attributes with NTBS values - currently all attributes takes only ULEB128 as values. In order to have attributes that takes NTBS values, new headers need to be defined.

The `.aeabi_subsection` directive should accept any subsection name, only checking the second and third argument if the name is known. This will allow adding new, possibly platform-specific subsections without needing to update the assembler to know about them, and is why the `.aeabi_subsection` directive has the `optional` and `type` arguments. This will then make it possible to emit attributes with NTBS values.

sivan-shani committed:
Add support for private subsections, add tests for type NTBS

sivan-shani committed:
fix string attr for elf + test

ostannard commented on llvm/include/llvm/Support/AArch64BuildAttributes.h:
```diff
@@ -0,0 +1,88 @@
+//
+// Build Attributes for the Arm® 64-bit Architecture (AArch64) 2024Q1
+//
+// https://github.com/ARM-software/abi-aa/blob/ada00cc8e04f421cce657e8d9f3a439c69437cf4/buildattr64/buildattr64.rst
```
I think this URL points to an un-committed PR, so I don't know if it will remain live after the PR is merged. It would be better to get the spec committed before landing this, but if that's not possible then I think it would be better to link to the PR here (https://github.com/ARM-software/abi-aa/pull/230/) because that URL will last longer.

ostannard commented on llvm/lib/Support/AArch64BuildAttributes.cpp:
```diff
@@ -0,0 +1,140 @@
+
+StringRef getVendorName(unsigned Vendor) {
+  switch (Vendor) {
+  case AEABI_FEATURE_AND_BITS:
```
These switch statements will need updating every time a new enum value is added, so I don't think the `VendorName` array is adding much, and you could just return string literals directly here.

sivan-shani replied:
The array is useful when referring to those string from elsewhere, avoiding possible typos.
e.g. in
`AArch64AsmPrinter::emitAttributes`
`AArch64AsmParser::parseDirectiveAeabiAArch64Attr`

ostannard commented on llvm/lib/Support/AArch64BuildAttributes.cpp:
```diff
@@ -0,0 +1,140 @@
+  }
+}
+VendorID getVendorID(StringRef Vendor) {
+  if (Vendor == VendorName[AEABI_FEATURE_AND_BITS]) {
```
For the string->enum functions, you can use `llvm::StringSwitch`: https://llvm.org/doxygen/classllvm_1_1StringSwitch.html

ostannard commented on llvm/lib/Support/AArch64BuildAttributes.cpp:
```diff
@@ -0,0 +1,140 @@
+    return TAG_PAUTH_SCHEMA;
+  return PAUTHABI_TAG_NOT_FOUND;
+}
+StringRef getPauthabiTagError() {
```
Unused.

ostannard commented on llvm/include/llvm/Support/AArch64BuildAttributes.h:
```diff
@@ -0,0 +1,88 @@
+
+namespace AArch64BuildAttributes {
+// AArch64 build attributes
+StringRef getSubsectionTag();
```
These are not going to change, so I think the indirection makes the code harder to read.

ostannard commented on llvm/lib/Support/AArch64BuildAttributes.cpp:
```diff
@@ -0,0 +1,140 @@
+    return TAG_FEATURE_GCS;
+  return FEATURE_AND_BITS_TAG_NOT_FOUND;
+}
+StringRef getFeatureAndBitsTagError() {
```
Unused.

ostannard commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -54,6 +54,7 @@
 #include "llvm/MC/MCStreamer.h"
 #include "llvm/MC/MCSymbol.h"
 #include "llvm/MC/TargetRegistry.h"
+#include "llvm/Support/ARMBuildAttributes.h"
```
What are we using from `ARMBuildAttributes.h` in this file, and can it be moved somewhere common?

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7806,6 +7815,261 @@ bool AArch64AsmParser::parseDirectiveSEHSaveAnyReg(SMLoc L, bool Paired,
+  AArch64BuildAttributes::SubsectionOptional IsOptional;
+  // options: optional/required
+  if (Parser.getTok().is(AsmToken::Identifier)) {
+    StringRef Optinality = Parser.getTok().getIdentifier();
```
Spelling: `Optionality`

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7806,6 +7815,261 @@ bool AArch64AsmParser::parseDirectiveSEHSaveAnyReg(SMLoc L, bool Paired,
+  Parser.Lex();
+  // Parsing finished, check for trailing tokens.
+  if (Parser.getTok().isNot(llvm::AsmToken::EndOfStatement)) {
+    Error(Parser.getTok().getLoc(), "unexpected token for AArch64 build "
```
Missing test.

sivan-shani replied:
test case added

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7806,6 +7815,261 @@ bool AArch64AsmParser::parseDirectiveSEHSaveAnyReg(SMLoc L, bool Paired,
+  std::unique_ptr<MCELFStreamer::AttributeSubSection> ActiveSubsection =
+      getTargetStreamer().getActiveAtributesSubsection();
+  if (nullptr == ActiveSubsection) {
+    Error(Parser.getTok().getLoc(),
```
Missing test.

sivan-shani replied:
test case added

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7806,6 +7815,261 @@ bool AArch64AsmParser::parseDirectiveSEHSaveAnyReg(SMLoc L, bool Paired,
+    if (AArch64BuildAttributes::NTBS == ActiveSubsectionType) {
+      Error(
+          Parser.getTok().getLoc(),
+          "active subsection type is NTBS (string), found ULEB128 (unsigned)");
```
Missing test.

sivan-shani replied:
test case added

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7806,6 +7815,261 @@ bool AArch64AsmParser::parseDirectiveSEHSaveAnyReg(SMLoc L, bool Paired,
+      TagStr != "") { // TagStr was a recognized string
+    if (0 != ValueInt && 1 != ValueInt) {
+      Error(Parser.getTok().getLoc(),
+            "Unknown AArch64 build attributes Value for Tag '" + TagStr +
```
Missing test.

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7806,6 +7815,261 @@ bool AArch64AsmParser::parseDirectiveSEHSaveAnyReg(SMLoc L, bool Paired,
+  // Parsing finished, check for trailing tokens.
+  if (Parser.getTok().isNot(llvm::AsmToken::EndOfStatement)) {
+    Error(Parser.getTok().getLoc(),
+          "unexpected token for AArch64 build attributes tag and value "
```
Missing test.

sivan-shani replied:
test case added

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp:
```diff
@@ -148,13 +150,177 @@ class AArch64TargetAsmStreamer : public AArch64TargetStreamer {
+        break;
+      case AArch64BuildAttributes::TAG_FEATURE_BTI:
+        OS << "\t." << AArch64BuildAttributes::getAttrTag() << "\t"
+           << AArch64BuildAttributes::getFeatureAndBitsTagsStr(
```
Since we've got a function to convert the enum value to a string, could these three cases be folded together?

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp:
```diff
@@ -148,13 +150,177 @@ class AArch64TargetAsmStreamer : public AArch64TargetStreamer {
+        AArch64TargetStreamer::emitAttribute(VendorName, Tag, Value, "",
+                                             Override);
+        break;
+      case AArch64BuildAttributes::TAG_PAUTH_PLATFORM:
```
Again, can these be folded together?

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64TargetStreamer.cpp:
```diff
@@ -151,3 +151,97 @@ llvm::createAArch64ObjectTargetStreamer(MCStreamer &S,
+      for (MCELFStreamer::AttributeItem &Item : SubSection.Content) {
+        if (Item.Tag == Tag) {
+          if (!Override) {
+            if ((unsigned(-1) != Value && Item.IntValue != Value) ||
```
Both sides of this `if` statement do the same thing (just slightly different wording in the assert).

sivan-shani replied:
The if/else capture 2 different cases, and gives an appropriate debug message accordingly.
The value out of it is:
1. proper debug message, which is helpful.
2. readability, clarifying why the return is happening.
Having this seems better then not, even though it is possible to reduce a bit of code.

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64TargetStreamer.cpp:
```diff
@@ -151,3 +151,97 @@ llvm::createAArch64ObjectTargetStreamer(MCStreamer &S,
+    }
+  }
+  assert(0 && "Can not add AArch64 build attribute: required subsection does "
+              "not exists");
```
Spelling: `exist`.

sivan-shani commented on llvm/test/MC/AArch64/aarch64-build-attributes-asm-private-subsections.s:
```diff
@@ -0,0 +1,50 @@
+.aeabi_subsection private_subsection_2, required, uleb128
+.aeabi_attribute 76, 257
+.aeabi_subsection private_subsection_3, optional, ntbs
+.aeabi_attribute 34, hello_llvm
```
Add support for quoted strings + test cases.

sivan-shani commented on llvm/test/MC/AArch64/aarch64-build-attributes-asm-err-attrs.s:
```diff
@@ -0,0 +1,62 @@
+.aeabi_subsection aeabi_feature_and_bits, optional, uleb128
+.aeabi_attribute Tag_PAuth_Platform, 1
+// ERR: error: Unknown AArch64 build attribute 'Tag_PAuth_Platform' for subsection 'aeabi_feature_and_bits' 
+// ERR-NEXT: Hint: options are: Tag_Feature_BTI, Tag_Feature_PAC, Tag_Feature_GCS
```
changing error message to match (removing 'hints')

sivan-shani committed:
fix spelling, add required test cases, remove unused, improve getting unsigned/string

sivan-shani commented:
Fixed according to comments above

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp:
```diff
@@ -148,13 +150,151 @@ class AArch64TargetAsmStreamer : public AArch64TargetStreamer {
+                                             Override);
+        break;
+      case AArch64BuildAttributes::TAG_FEATURE_BTI:
+        LLVM_FALLTHROUGH;
```
You don't need to use `LLVM_FALLTHROUGH` if there is no code between two `case` statements.

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64ELFStreamer.cpp:
```diff
@@ -148,13 +150,151 @@ class AArch64TargetAsmStreamer : public AArch64TargetStreamer {
+    case AArch64BuildAttributes::VENDOR_UNKNOWN: {
+      // Keep the data structure consistent with the case of ELF emission
+      // (important for llvm-mc asm parsing)
+      OS << "\t" << SubsectionTag << "\t" << SubsectionName << ", "
```
This is the same in each case, so can be moved outside the `switch`, leaving just the assertions here.

sivan-shani commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7806,6 +7815,267 @@ bool AArch64AsmParser::parseDirectiveSEHSaveAnyReg(SMLoc L, bool Paired,
+                Optionality);
+      return true;
+    }
+    if (HasActiveSubsection &&
```
fixed + test added

sivan-shani commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7806,6 +7815,267 @@ bool AArch64AsmParser::parseDirectiveSEHSaveAnyReg(SMLoc L, bool Paired,
+      return true;
+    }
+    if (HasActiveSubsection &&
+        (SubsectionName == ActiveSubsection->VendorName)) {
```
fixed + test added

sivan-shani committed:
fix according to comments

sivan-shani commented:
Fixed according to comments above

sivan-shani committed:
Add AArch64BuildAttributes.cpp to lib/Support/BUILD.gn

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64TargetStreamer.cpp:
```diff
@@ -182,6 +182,16 @@ AArch64TargetStreamer::getActiveAtributesSubsection() {
 }
 
+std::unique_ptr<MCELFStreamer::AttributeSubSection>
+AArch64TargetStreamer::getActiveSubsectionByName(StringRef name) {
```
This name is misleading, because this function doesn't check if the subsection is active. This should also have "Attribute" in the name, because "subsection" has other meanings.

sivan-shani replied:
between `getAtributeSubsectionByName` and `getAtributesSubsectionByName`, opt for the later.

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7846,6 +7839,13 @@ bool AArch64AsmParser::parseDirectiveAeabiSubSectionHeader(SMLoc L) {
     return true;
   }
 
+  bool SubsectionExists = true;
```
You don't need this extra `bool`, you can just check `if (ExistingSubsection)` below.

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7858,20 +7858,19 @@ bool AArch64AsmParser::parseDirectiveAeabiSubSectionHeader(SMLoc L) {
                   "' already exists with optionality defined as '" +
-                  Twine(ActiveSubsection->IsOptional) + "' and not '" +
+                  Twine(ExistingSubsection->IsOptional) + "' and not '" +
                   Twine(IsOptional) + "' (0: required, 1: optional)");
```
I think the numbers here are only used in the ELF encoding, so they aren't relevant to the assembly programmer. The diagnostic should refer to the identifiers used in the assembly.

sivan-shani replied:
fixed, also for type.

sivan-shani committed:
fix according to comments above

sivan-shani commented:
Fixed according to comments above

sivan-shani committed:
fix error in a test file

ostannard reviewed (approved):
LGTM.

nico commented:
Seems seems to break tests: http://45.33.8.238/linux/158130/step_11.txt

Please take a look, and revert for now if it takes a while to fix.

antmox commented:
Hello, could this patch cause this failure ? :
https://lab.llvm.org/buildbot/#/builders/65/builds/11046

sivan-shani commented:
@nico @antmox the very same 5 .ll tests passed upstream pre-merge
https://buildkite.com/llvm-project/github-pull-requests/builds/139398#01948e1c-bd4f-41d6-878b-789113e2cbea
Not sure how to understand the later failure. 
Still investigating.

sivan-shani commented:
What might be a way to figure the diff between the pipelines (the pre-merge that those same tests pass on) and the clang-aarch64-quick that fail, regarding build command, environment, etc?

sivan-shani commented:
Since all tests passed at pre-merge and rest of relevant tests passing on all pipelines, it seems as the issue is not the patch at large but those specific 5 .ll tests. Will disable those specific 5 tests for now to clear the pipeline.

kazutakahirata commented:
> Since all tests passed at pre-merge and rest of relevant tests passing on all pipelines, it seems as the issue is not the patch at large but those specific 5 .ll tests. Will disable those specific 5 tests for now to clear the pipeline.

I've reverted the patch.  I've reproduced the test failures in my build tree also.  I'm happy to try your new version whenever it's ready.  Thanks!

sivan-shani commented:
Yes, I realized that it is actually failing not only on AArch64 but also on x86 and was about to revert it myself.
Thank you, working on a fix.

MaskRay commented on llvm/lib/Support/AArch64BuildAttributes.cpp:
```diff
@@ -0,0 +1,117 @@
+#include "llvm/ADT/StringSwitch.h"
+
+namespace llvm {
+namespace AArch64BuildAttributes {
```
In ARM we have ARMBuildAttrs. Should this be shortened to AArch64BuildAttrs?

sivan-shani replied:
That indeed will be more consistent, on the other hand, Attributes is more descriptive than Attrs, so perhaps changing ARMBuildAttrs to ARMBuildAttributes instead will give both the benefit of consistency and of clarity?

Other files use Attributes as well, and not Attrs, for example: ELFAttributes.cpp, ELFAttributeParser.cpp,  HexagonAttributes.cpp, HexagonAttributeParser.cpp, CSKYAttributes.cpp, CSKYAttributeParser.cpp, MSP430Attributes.cpp, MSP430AttributeParser.cpp
